backend/src/main/java/ai/giskard/service/ModelService.java

```diff
@@ -37,74 +42,121 @@ public class ModelService {
+    private MLWorkerWSRunModelForDataFrameDTO getRunModelForDataFrameResponse(ProjectModel model, Dataset dataset, Map<String, String> features) {
+        // Initialize the parameters and build the Data Frame
+        MLWorkerWSDataFrameDTO dataframe = MLWorkerWSDataFrameDTO.builder()
+            .rows(
+                List.of(
+                    MLWorkerWSDataRowDTO.builder().columns(
                         features.entrySet().stream()
-                            .filter(entry -> !shouldDrop(dataset.getColumnDtypes().get(entry.getKey()), entry.getValue()))
-                            .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue))
-                    ))
-                    .build()
-            );
+                            .filter(entry -> !shouldDrop(
+                                dataset.getColumnDtypes().get(entry.getKey()),
+                                entry.getValue()
+                            )).collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue))
+                    ).build()
+                )
+            ).build();
+
+        MLWorkerWSRunModelForDataFrameParamDTO param = MLWorkerWSRunModelForDataFrameParamDTO.builder()
+            .model(MLWorkerWSArtifactRefDTO.fromModel(model))
+            .dataframe(dataframe)
+            .build();
+
         if (dataset.getTarget() != null) {
-            requestBuilder.setTarget(dataset.getTarget());
+            param.setTarget(dataset.getTarget());
         }
         if (dataset.getColumnTypes() != null) {
-            requestBuilder.putAllColumnTypes(Maps.transformValues(dataset.getColumnTypes(), ColumnType::getName));
+            param.setColumnTypes(Maps.transformValues(dataset.getColumnTypes(), ColumnType::getName));
         }
         if (dataset.getColumnDtypes() != null) {
-            requestBuilder.putAllColumnDtypes(dataset.getColumnDtypes());
+            param.setColumnDtypes(dataset.getColumnDtypes());
         }
-        response = client.getBlockingStub().runModelForDataFrame(requestBuilder.build());
-        return response;
+
+        // Perform the runModelForDataFrame action and parse the reply
+        MLWorkerWSBaseDTO result = null;
+        try {
+            result = mlWorkerWSCommService.performAction(
+                model.getProject().isUsingInternalWorker() ? MLWorkerID.INTERNAL : MLWorkerID.EXTERNAL,
+                MLWorkerWSAction.runModelForDataFrame,
+                param
+            );
+        } catch (JsonProcessingException e) {
+            throw new RuntimeException(e);
+        } catch (NullPointerException e) {
+            throw new NullPointerException("Did not get an validate ML Worker RunModelForDataFrame reply");
+        }
+        if (result instanceof MLWorkerWSRunModelForDataFrameDTO response) {
+            return response;
+        }
+        throw new NullPointerException("Cannot get ML Worker RunModelForDataFrame reply");
```

Comment:
I don't get this NPE usage (and also in other places where you added it). There's no null pointer in this context, you're just reacting to an unexpected result. You should create a subclass of `RuntimeException` and throw it if needed.